### PR TITLE
[FIX] Specify marker file names for KIT data (MEG)

### LINKED ISSUE #45: naming multiple marker files for KIT MEG systems
Currently the specification only has details on how to name a single marker file associated with KIT MEG systems ([see here](https://github.com/bids-standard/bids-specification/blob/master/src/99-appendices/06-meg-file-formats.md#kityokogawaricoh)).
MNE-python however is able to handle multiple .mrk files and averages them if more than one is provided.
I wrote a PR to export the .mrk files with MNE-BIDS (https://github.com/mne-tools/mne-bids/pull/114) but since we have no specification on what to call the files if multiple are used this PR cannot continue until this issue has been resolved.

While the majority of the time it seems only one .mrk file is ever used with any given .con file, sometimes the researchers will do a `-pre` and `-post` recording .mrk file. To this end I propose possibly having a `-pre` and `-post` tag after the `_markers` tag in the file name.
Eg.
```
sub-control01_ses-001_task-rest_run-01_markers[-pre|-post].<mrk,sqd>
```
Not sure if this is the best way to do it, so I am very open to better suggestions.
@teonbrooks

### PR DESCRIPTION
**Pull request by monkeyman192**
closes #45

### CODE REVIEW

**Comment by jasmainak** [comment]
Looks reasonable to me

**Comment by chrisfilo** [comment]
Thank you for opening the PR.

I got one concern: you are using the `acq` filename keyword which is free form everywhere else but in this particular case you only two values are allowed. This makes the spec inconsistent. I would suggest proposing another keyword.

**Comment by monkeyman192** [comment]
I just did this because it's what @robertoostenveld and @teonbrooks recommended.
I agree however that there is the potential of conflict by using `acq`, hence why I initially suggested to just suffix `marker` with `pre` or `post`.
Essentially making the line
```
sub-control01_ses-001_task-rest_run-01_markers[-pre,-post].<mrk,sqd>
```

**Comment by chrisfilo** [comment]
I would prefer new tag (like `marker`) instead of reusing `acq` for reasons stated above, but we need more MEG people to join the conversation.

**Comment by teonbrooks** [comment]
since `_markers` is only specified for KIT in the spec, i think we could generalize it to a keyword as @chrisfilo suggested. I think having `-pre` and `-post` as its possible values is fine.

**Comment by jasmainak** [comment]
What would be the suffix if the key-value pair is `marker-pre` and `marker-post` ? Shouldn't all files in BIDS follow the format:

``<key1>-<value1>_<key2>-<value2>_..._<suffix>.<extension>``

?

**Comment by chrisfilo** [comment]
This is a good question. There should be a prefix.

**Comment by monkeyman192** [comment]
Could just reuse `meg`?
although it would then not make sense for when you don't specify `pre` or `post`...

**Comment by jasmainak** [comment]
I think the most backwards compatible solution would be to leave the suffix as it is (`_markers.mrk`) and instead come up with a new key-value pair to indicate weather it's recorded before or after a trial.

**Comment by monkeyman192** [comment]
so I was looking through the specification for something else and noticed that the `recording` kvp is completely unused for MEG as far as I can tell, so maybe we could use this?
More specifically the name of the file would become:
`sub-<label>[_ses-<label>][_task-<label>][_run-<label>][_recording-<pre,post>]_markers.<mrk,sqd>`
Also, I think limiting this to two files is fine because you can't have more than two anyway I think.

**Comment by chrisfilo** [comment]
I think `recording` is already used in physiological data: https://github.com/bids-standard/bids-specification/blob/master/src/04-modality-specific-files/04-physiological-and-other-continous-recordings.md#physiological-and-other-continuous-recordings

**Comment by monkeyman192** [comment]
would it matter if were used for a different context for a pretty niche case?
I don't think it would cause any conflict as `recording` isn't used in any other context for MEG data

**Comment by chrisfilo** [comment]
I would prefer to avoid such internal inconsistency by picking a different keyword.

**Comment by jasmainak** [comment]
@chrisfilo what would you suggest that maintains internal consistency? I suppose trying to use any of the existing keys would lead to consistency issues, so adding a new key-value pair seems to be the only solution.

**Comment by chrisfilo** [comment]
Yes. I think adding a new keyword instead of redefining an old one would improve the consistency and make the standard more intuitive for people working with multiple modalities.

**Comment by monkeyman192** [comment]
I completely see the justification for wanting to introduce another kvp to avoid inconsistencies, however it just seems silly to add another for just one single case.
I am not *against* adding another kvp, I just think we keep adding synonyms for recording (`run`, `acquisition`), sometimes it may be *more* confusing having more terms. When someone looks at the specification they may wonder why there are so many terms that all mean essentially the same thing.

Finally do you have a suggestion for an appropriate key? I can't really think of one that would be good. If we don't care about having the kvp used elsewhere then we could just use `mark-<pre,post>_marker` even if it sort of doubles up the `marker` tag.

**Comment by robertoostenveld** [comment]
I propose to consider this also in relation to EEG and especially iEEG specifications. For the iEEG it is  not uncommon to have pre and post-operative data. For EEG it would be possible (but not common) to have pre- and post-measurement electrode digitizations.

**Comment by robertoostenveld** [comment]
besides the two `pre` and `post`, I think that more than two files with auxiliary data (i.e. data that enriches the core _meg/_eeg/_ieeg dataset) would also be possible. The labels `pre` and `post` are just two possible values. E.g. consider the case that multiple photos are taken of EEG electrodes or MEG markers, like with [photogrammetry](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5432580/). The series of photos would be from different angles, but not from different moments per see.

**Comment by monkeyman192** [comment]
maybe having `meas` (for `measurement`) as the `key` could work?
For the marker measurements we could say that the values have to be one of `pre` or `post`.
For other cases such as the photogrammetry the values could be un-restricted.

**Comment by monkeyman192** [comment]
Any further thoughts on this? It would be quite nice to get this sorted as it blocks a few things that I am working on and the sooner it is sorted the less impact it has on all the data converted to BIDS format by where I work.

**Comment by robertoostenveld** [comment]
> What would be the suffix if the key-value pair is `marker-pre` and `marker-post` ? Shouldn't all filenames in BIDS follow the format:

The suffix is already specified as headshape, i.e. as
`sub-<label>[_ses-<label>][_acq-<label>]_headshape.<manufacturer_specific_extension>`

In the thread above `marker` is suggested as suffix. I don't see the need for that, and I don't see it anywhere in the current specification.

Re: restricted versus free form values: I think that just `pre` and `post` applies often, but will not enough in practice. Therefore I think that the value should free form and hence `acq` as the key would work.  I have a general preference of not introducing too many keys, especially not if they conceptually have the same meaning.

**Comment by jasmainak** [comment]
I agree with the views shared above by @robertoostenveld and @sappelhoff . We should reuse the `acq` key

**Comment by monkeyman192** [comment]
ok, I have updated the PR to fix the merge conflict and update the text to reflect the above discussion

**Comment by chrisfilo** [comment]
From a perspective of someone writing software that will ingest such datasets the current wording seems to liberal. For example, it would be completely ok for someone to use `acq-pre` for acquisition after  MEG or to encode something completely different. To make this semantic usable for data consumers I think you should go to semi-controlled vocabulary where `post` and `pre` are reserved keywords that have a singular meaning.

**Comment by robertoostenveld** [comment]
why do you consider it required to be (semi) controlled here, and not for the initial use of the `_acq` label. For reference, this is what the specification contains: 

> The OPTIONAL acq-<label> key/value pair corresponds to a custom label the user MAY use to distinguish a different set of parameters used for acquiring the same modality. For example this should be used when a study includes two T1w images - one full brain low resolution and and one restricted field of view but high resolution. In such case two files could have the following names: sub-01_acq-highres_T1w.nii.gz and sub-01_acq-lowres_T1w.nii.gz, however the user is free to choose any other label than highres and lowres as long as they are consistent across subjects and sessions. In case different sequences are used to record the same modality (e.g. RARE and FLASH for T1w) this field can also be used to make that distinction. At what level of detail to make the distinction (e.g. just between RARE and FLASH, or between RARE, FLASH, and FLASHsubsampled) remains at the discretion of the researcher.

In the use for MEG headshape recordings, I would also say that " the user is free to choose any other label than pre and post as long as they are consistent across subjects and sessions".

**Comment by chrisfilo** [comment]
Hi @robertoostenveld,

To be honest I would prefer is a different filename keyword was used for this purpose (see my comments above) for the same reason you mentioned - consistency with the usage of `acq` in other contexts. I suggested this solution as a compromise seeing the opposition towards adding new keywords.

As for the suggestion to "let users choose any label they want as long as it is consistent". I think this is what users would do with the current spec. The addition of a more strict controlled vocabulary enables programmatic uses. A piece of software could in such situation automatically find which files recording was made before MEG without having to ask the user for corresponding `acq` label.

**Comment by monkeyman192** [comment]
@chrisfilo what keyword would you suggest be used instead of `acq`?
I agree it is good to have consistency, but it seems like most others are pretty happy with using `acq`.
Also, @robertoostenveld I somewhat disagree with
> I think that just pre and post applies often, but will not enough in practice.

What other situations do you envisage?
In terms of the recording process where I work there will only ever be at most 2 markers associated with any one .con file, one before and one afterwards.

**Comment by yarikoptic** [comment]
Although I am typically the one demanding semantically clear field, here `acq` seems to fit the bill, especially if we keep in mind that we are taking about  `_markers`.
If someone needs programmatic treatment of those files, I think it would be ok to demand specific pair of values, defaulting to pre and post. Specification could note that those are suggested values to describe pre/post data collections

File: src/99-appendices/06-meg-file-formats.md [hunk 1]
```diff
@@ -180,10 +180,17 @@ Where:
 
 ```Text
 sub-control01_ses-001_task-rest_run-01_meg/
-    sub-control01_ses-001_task-rest_run-01_markers.<mrk,sqd>
+    sub-control01_ses-001_task-rest_run-01[_acq-<acq-label>]_markers.<mrk,sqd>
```
Comment (yarikoptic) [inline src/99-appendices/06-meg-file-formats.md]: Isn't there also a table to adjust in the supplementals for having acq allowed here? Sorry, hard to lookup on the phone atm

Comment (monkeyman192) [reply src/99-appendices/06-meg-file-formats.md]: yes! Good call! I forgot about that table.

**Comment by chrisfilo** [comment]
> @chrisfilo what keyword would you suggest be used instead of `acq`?

A new one as we discussed before. For example `megrel` (as in relation to MEG acquisition - `pre` or `post`). I am not hell bent on this though.

My main point is `pre` and `post` values need to have specific definitions and data curators should be required to use those values only and only if the cases described in the definitions apply. If we only suggest. If you really want to use `acq` that is ok, but in such case the following will apply:

a) For non `_markers` files curators can use any value of `acq` for any type of acqusiton
b) For `_markers` files curators can you any value of `acq` for any type of acqusiton as long as the values are not `pre` or `post`. `pre` or `post` can only be used to denote data acquisition before or after an MEG recording

Alternatively we can abandon the attempt to standardize (make it machine accessible) this aspect of metadata and just say that if you have two `_markers` files: one before and one after MEG recording you can use `acq` filename key with wahtever values you prefer to distinguish them.

**Comment by yarikoptic** [comment]
How is it important to know the meaning/semantic for pre/post? 
 from #45 summary it seems that pre/post is just for cases when indeed multiple digitizations were done (so not even always), and then they likely just get averaged (so no semantic actually needed). So it sounds like too much battle for no winner ;-)

**Comment by jasmainak** [comment]
yes, exactly my question too. In MNE, they are averaged anyway when reading the files. There are some cases according to @monkeyman192 when they can be analyzed separately but I would say it's the 20% rather than 80% use case. And at this point, we are probably bikeshedding :) 

For mne-bids, it's probably easier to allow acq to be free form. Otherwise, the user has to specify the pre and post labels manually. And the more work you give people for converting datasets, the less likely they will be to adapt the standard ...

**Comment by monkeyman192** [comment]
> Otherwise, the user has to specify the pre and post labels manually.

Not really. See https://github.com/mne-tools/mne-bids/pull/173

**Comment by robertoostenveld** [comment]
Re "megrel" as name. The repeated acquisition of geometrical data (as the markers for the KIT system in this example) also happens in other contexts. E.g. repeated CTs to localize iEEG electrodes, repeated electrode digitizations using a Polhemus, repeated photo's for photogrammetry. Since this is all related to secondary data (the main data type is anat/meg/eeg/ieeg) it does not pertain to "runs" (as keyword). But it is also not MEG specific.

Re "What other situations do you envisage?" For example repeated measurements of the markers prior to an experiment (so pre1, pre2, pre3). Or measurements between runs (in case the subject comes out of the MEG system and/or coils that need to be reattached). 

I hope that we can find a nomenclature that can scale further than just "KIT markers", but more broadly meg and other modalities. We already have

    sub-0001_ses-01_acq-photo1_photo.jpg
    sub-0001_ses-01_acq-photo2_photo.jpg

in https://github.com/bids-standard/bids-specification/blob/master/src/04-modality-specific-files/04-intracranial-electroencephalography.md for me it feels natural to use that here as well. Please search for `acq-` on that page to see the different use cases.

**Comment by chrisfilo** [comment]
I think I understand the context better now. Thank you.

You don't want to standardize how post/pre MEG marker files should be named. You just want to reiterate in the spec that `acq` can be used for avoiding name collisions for all sort of acquisition differences including relation to MEG. I think that is perfectly fine.

In such a case, I would rephrase the line https://github.com/bids-standard/bids-specification/pull/62/files#diff-56231ce79435a3bc626ef265c47d4acdR190 to include more broad language and treat `post`, `pre` as examples rather than recommendations.

**Comment by monkeyman192** [comment]
> In such a case, I would rephrase the line https://github.com/bids-standard/bids-specification/pull/62/files#diff-56231ce79435a3bc626ef265c47d4acdR190 to include more broad language and treat post, pre as examples rather than recommendations.

I'll try reword it to be more broad, but I still think that `pre` and `post` should be the recommended values.
But a recommendation is just that. People don't *need* to abide by it, but it would be nicer if they did.

File: src/99-appendices/06-meg-file-formats.md (hunk 1 above)
Comment (monkeyman192) [reply src/99-appendices/06-meg-file-formats.md]: Actually, I think looking at the table it is already included. `acq` is already an optional parameter for `meg`, and there is probably no point in added `marker` to the top column title since it is only for a single manufacturer...

Comment (chrisfilo) [reply src/99-appendices/06-meg-file-formats.md]: it might be worth adding this for the sake of completeness

File: src/99-appendices/01-contributors.md [hunk 2]
```diff
@@ -100,6 +100,7 @@ your name is not listed, please add it.
 -   Jean-Baptiste Poline 📖📢🤔🎨
 -   Vasudev Raguram 💻🎨📖🔧
 -   Ariel Rokem 📖
+-   Matt Sanderson 📖💻
```
Comment (chrisfilo) [inline src/99-appendices/01-contributors.md]: Please also add your github username to https://github.com/bids-standard/bids-specification/blob/master/CODEOWNERS

File: src/99-appendices/06-meg-file-formats.md [hunk 3]
```diff
@@ -180,10 +180,19 @@ Where:
...
```
Comment (yarikoptic) [inline src/99-appendices/06-meg-file-formats.md]: here and elsewhere, should be `_acq-<label>` as per my unification through the entire specification a little back in #93

**Comment by robertoostenveld** [comment]
Especially in the KIT case, to which this initially relates

**Comment by chrisfilo** [comment]
+1

File: src/99-appendices/06-meg-file-formats.md [hunk 4]
```diff
@@ -180,10 +180,18 @@ Where:
 
 ```Text
 sub-control01_ses-001_task-rest_run-01_meg/
-    sub-control01_ses-001_task-rest_run-01_markers.<mrk,sqd>
+    sub-control01_ses-001_task-rest_run-01[_acq-<label>]_markers.<mrk,sqd>
     sub-control01_ses-001_task-rest_run-01_meg.<con,sqd>
 ```
 
+Marker coil files MUST have an acquisition parameter if there is more than
```
Comment (robertoostenveld) [inline src/99-appendices/06-meg-file-formats.md]: I propose to write it as 

If there are multiple files with marker coils, they must have the `acq-<label>` parameter and no more that two files may be associated with one raw data file. While the acquisition parameter can take any value, it is RECOMMENDED that if the two marker measurements occur before and after the raw data acquisition,
 `pre` and `post` are used to differentiate the two situations.

File: src/99-appendices/06-meg-file-formats.md [hunk 5]
```diff
@@ -156,11 +156,11 @@ More about the 4D neuroimaging/BTi data organization at:
...
-`.con`) file with its associated marker coil file (`.mrk`), which contains coil
-positions in the acquisition system’s native space. Head points and marker
-points in head space are acquired using third-party hardware. One SHOULD
-rename/create a father run specific directory and keep the original files for
-each run inside.
+`.con`) file with its associated marker coil file(s) (`.sqd`, `.mrk`), which
+contains coil positions in the acquisition system’s native space. Head points
+and marker points in head space are acquired using third-party hardware. One
+SHOULD rename/create a further run specific directory and keep the original
+files for each run inside.
```
Comment (sappelhoff) [inline src/99-appendices/06-meg-file-formats.md]: this is colliding with #19 to some extend. Could you review that PR @monkeyman192?

Comment (monkeyman192) [reply src/99-appendices/06-meg-file-formats.md]: I'll wait for that PR to go through then rebase this onto the master to pick up the changes with the suggested rewording by @robertoostenveld

**Comment by monkeyman192** [comment]
@sappelhoff ok, rebased and hopefully addressed your comment as well as addressing @yarikoptic 's comment and @robertoostenveld 's comment (I did reword it a bit however as I think the wording wasn't quite right. Hopefully you agree!!)

File: src/99-appendices/06-meg-file-formats.md [hunk 6]
```diff
@@ -173,10 +173,17 @@ sub-control01/
             sub-control01_ses-001_task-rest_run-01_meg
             sub-control01_ses-001_task-rest_run-01_meg.json
             sub-control01_ses-001_task-rest_run-01_channels.tsv
-            sub-control01_ses-001_task-rest_run-01_markers.<mrk,sqd>
+            sub-control01_ses-001_task-rest_run-01[_acq-<label>]_markers.<mrk,sqd>
```
Comment (sappelhoff) [inline src/99-appendices/06-meg-file-formats.md]: Actually, according to the text in the [entities table appendix](https://bids-specification.readthedocs.io/en/stable/99-appendices/04-entity-table.html), the `acq` label must come before the `run` label :thinking: 

Could you change that @monkeyman192 ? (Sorry but I found it just now ...)

And it's another case where the validator is a bit too permissive, because when I tested it, it passed

Comment (sappelhoff) [reply src/99-appendices/06-meg-file-formats.md]:
```suggestion
            sub-control01_ses-001_task-rest[_acq-<label>]_run-01_markers.<mrk,sqd>
```

Comment (monkeyman192) [reply src/99-appendices/06-meg-file-formats.md]: yep, sorry, been sidetracked! Will get onto it asap!

**Comment by effigies** [comment]
@jasmainak 
>  It would be nice if the CIs also build the doc and show it in the artifacts tab

https://496-150465237-gh.circle-artifacts.com/0/home/circleci/project/site/99-appendices/04-entity-table.html

**Comment by jasmainak** [comment]
Oh nice, sorry missed it. Thanks! And PR looks good to me.

**Comment by sappelhoff** [comment]
thanks @monkeyman192
